internal/handlers/handler_firstfactor.go

```diff
@@ -2,127 +2,200 @@ package handlers
 
 import (
 	"fmt"
+	"math/rand"
 	"time"
 
 	"github.com/authelia/authelia/internal/authentication"
+	"github.com/authelia/authelia/internal/configuration/schema"
+	"github.com/authelia/authelia/internal/logging"
 	"github.com/authelia/authelia/internal/middlewares"
 	"github.com/authelia/authelia/internal/regulation"
 	"github.com/authelia/authelia/internal/session"
+	"github.com/authelia/authelia/internal/utils"
 )
 
-// FirstFactorPost is the handler performing the first factory.
-//nolint:gocyclo // TODO: Consider refactoring time permitting.
-func FirstFactorPost(ctx *middlewares.AutheliaCtx) {
-	bodyJSON := firstFactorRequestBody{}
-	err := ctx.ParseBody(&bodyJSON)
+func getDelayAuthSettings(config schema.AuthenticationBackendConfiguration) (bool, time.Duration) {
+	log := logging.Logger()
+
+	if !config.DisableDelayAuth {
+		var duration time.Duration
+
+		rand.Seed(time.Now().UnixNano())
+
+		if config.File != nil {
+			algorithm, err := authentication.ConfigAlgoToCryptoAlgo(config.File.Password.Algorithm)
+			if err != nil {
+				panic(err)
+			}
+
+			password := utils.RandomString(20, authentication.HashingPossibleSaltCharacters)
+			start := time.Now()
+			_, _ = authentication.HashPassword(password, "",
+				algorithm, config.File.Password.Iterations,
+				config.File.Password.Memory*1024, config.File.Password.Parallelism,
+				config.File.Password.KeyLength, config.File.Password.SaltLength)
+			duration = time.Since(start) + (time.Duration(rand.Int31n(50) + 150))
+		}
+
+		if duration < 450*time.Millisecond {
```

Comment:
The problem I see with the constant at 450ms is that if the app network is overloaded for instance the latencies of LDAP queries will increase and therefore there will be no delay anymore. If there is no delay the mitigation does not work anymore. It's pretty much the same issue as in my proposal with the threshold being exceeded.